### PR TITLE
[Enhancement] Making the focus state of button text in Button block more visible

### LINKED ISSUE #15051: Button Block: Focus style for the button text is not clear
As noted by @afercia in #10128, the focus style of button text within the button block is not clear. In this GIF, you can see a _very_ subtle cursor sitting in between the `t` and `e` of the placeholder text: 

button-focus

This does not make it entirely clear that your focus is within the button, and that typing will add text to this area. 

When you start typing, this blinking cursor is a little more evident: 

button-focus-text

At a minimum, we should try to improve the visibility of the blinking cursor in the placeholder state, but I'd recommend we go further than that and sort out a stronger indicator of focus for the button as well.

### PR DESCRIPTION
**Pull request by nfmohit**
## Description
This PR tries to close #15051 which reports the focus style of the button text within the button block not being clear.

The placeholder on this text input is essentially an overlay. It has a reduced opacity set so that the cursor shows up. But naturally, it barely does. This PR just reduces the opacity further only when the input is focused so that the cursor, i.e. the focus state, is more visible.

I don't think it is the most ideal approach to this, but at least it's a start. I'd request some feedback and suggestions on how we could improve it. Thanks ❤️ 

## How has this been tested?
This PR has been tested by going through the following steps:

1. Started a new post using the Gutenberg editor.
2. Added the "Button" block.
3. Focused on the button text input.
4. Made sure the opacity of the placeholder reduces making the cursor more visible.

## Types of changes
This PR just simply reduces the opacity of the placeholder on `:focus` state of the text input.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by kjellr** [comment]
Thank you, @nfmohit-wpmudev! This is a definite improvement. The previous layering of these elements is definitely a bug in retrospect. 

Upon trying this out, I wonder if we can go a little bit further with two changes before merging this: 

1. Other placeholder text we have disappears completely on select. I think we should adopt that here: 

typing

2. It's confusing that the cursor is in the center of the block here, since that's not actually where the text starts when you type. Can we move the cursor to the beginning when there's no text?

Here's what that would work out to: 

button-typing-new

Thank you for taking this on so quickly.

**Comment by nfmohit-wpmudev** [comment]
Thank you so very much for your feedback @kjellr ❤️ I have pushed some commits that address your suggestions.

**Comment by nfmohit-wpmudev** [comment]
Thank you very much for the quick review @kjellr ❤️ 

Sincere apologies if I'm missing something but the cursor doesn't stay active for me when I change the alignment of the button. I think that maybe because the focus state on the button is lost when I click elsewhere. Here's a screencast:

**Comment by nfmohit-wpmudev** [comment]
There's one other thing I could do - remove the `keepPlaceholderOnFocus` attribute from the `RichText` component so that after focus, the placeholder is removed completely. I'm not sure if this is an appropriate behavior but here's how it'd look:



Let me know if you like it.

**Comment by kjellr** [comment]
> Sincere apologies if I'm missing something but the cursor doesn't stay active for me when I change the alignment of the button. I think that maybe because the focus state on the button is lost when I click elsewhere.

Interesting! This only seems to happen in Safari now that I'm testing more widely. Chrome and Firefox lose focus when I click a new alignment. 🤔 I'll do a little digging tomorrow and see if I can come up with a decent solution.

File: packages/block-library/src/button/editor.scss
Comment (kjellr) [inline packages/block-library/src/button/editor.scss]: @nfmohit-wpmudev I think left-aligning the cursor here at all times (not just on focus) seems like a reasonable fix for [the Safari issue I described above](https://github.com/WordPress/gutenberg/pull/15058#issuecomment-484944031). 

With that change, the cursor stays active in the field after switching alignment, but it's left-aligned instead of centered all of a sudden. Since the field isn't technically focused, the placeholder text is active again (and the cursor is therefore pretty dim), but I'm not sure there's an easy way to fit that part. 

Here's a quick GIF: 

animation

Comment (nfmohit-wpmudev) [reply packages/block-library/src/button/editor.scss]: > @nfmohit-wpmudev I think left-aligning the cursor here at all times (not just on focus) seems like a reasonable fix for the Safari issue I described above.
I have applied this suggestion on [this commit](https://github.com/WordPress/gutenberg/pull/15058/commits/a06f391f648fec87ceb7e4c6172bc86a42041cfe).

In the case of Safari, I'm really not sure how to retain the focus styles when alignment is changed without affecting the other browsers, as the focus state itself is not retained physically. I'm looking forward to feedback and suggestions regarding this. Thank you ❤️

Comment (kjellr) [reply packages/block-library/src/button/editor.scss]: Yeah, I ran into that issue too. It's a bit of an edge case, so I think this is good enough for now. 👍 Nice work!

**Review by kjellr — APPROVED** [review]
This looks good on my end! The placeholder now behaves like I'd expect it to. Here's a GIF: 

animation

I'd like to just get one last accessibility review before merge. @afercia, since does this seem like a solid improvement from your end? Thank you!

**Comment by afercia** [comment]
Fixing the caret is a good start, thank you! However, in the editing context the button is equivalent to an input field: it should have aso an additional focus indicator, for example an outline, like all the input fields.

**Review by kjellr — CHANGES_REQUESTED** [review]
Thanks, @afercia! 

@nfmohit-wpmudev, let's add a border like this so it has a more visible focus state: 

focus-state

Something like this should take care of it: 

```
// Add a blue border to indicate focus. 
.block-editor-rich-text__editable:focus {
	box-shadow: 0 0 0 1px $white, 0 0 0 3px $blue-medium-500;

	// Windows High Contrast mode will show this outline, but not the box-shadow.
	outline: 2px solid transparent;
	outline-offset: -2px;
}
```

For now, we'll also need to move the `.block-library-button__inline-link` down by `2px` in order to make sure this border [doesn't overlap](https://cloudup.com/cLNoA65vdxP) the URL field. That'll eventually be unnecessary once #10128 is merged in. 

Once that's in place, this should be good to 🚢!

**Comment by nfmohit-wpmudev** [comment]
Thank you for your feedback @afercia and thank you so very much for your suggestions @kjellr ❤️ I have addressed them in [this commit](https://github.com/WordPress/gutenberg/pull/15058/commits/1bd0e74d44b6c47c944bcaae9f8fc9a6f62b3688).

**Review by kjellr — APPROVED** [review]

**Comment by kjellr** [comment]
Looks great, thanks @nfmohit-wpmudev. Should be good to merge after the tests pass.

**Comment by nfmohit-wpmudev** [comment]
Thank you very much @kjellr ❤️

**Pull request merged**
